Exclude OperatingSystemMXBean PhysicalMemory test in Docker

pshipton opened the pull request:
Part of tests testOSMXBeanLocal and testOSMXBeanRemote, which run
org.openj9.test.management.TestOperatingSystemMXBean.runTestOSMXBean()

Issue https://github.com/eclipse/openj9/issues/12038

pshipton commented:
I'll add plinux.

pshipton commented:
Even better would be to detect the VM is running in docker, not sure how to do that.

keithc-ca commented on test/functional/JLM_Tests/src/org/openj9/test/management/TestOperatingSystemMXBean.java:
```diff
@@ -63,6 +63,21 @@
 	private static final int NTHREADS = 1;
 	private static final int NITERATIONS = 10;
+	
+	static boolean isPhysicalMemoryTestExcluded() {
+		/* Exclude on Linux aarch64 and x86-64
+		 * https://github.com/eclipse/openj9/issues/12038
+		 */
+		String osname = System.getProperty("os.name");
+		if (!("Linux").equals(osname)) {
+			return false;
+		}
+		String arch = System.getProperty("os.arch");
+		if (!"amd64".equals(arch) && !"aarch64".equals(arch)) {
+			return false;
+		}
+		return true;
+	}
```
This can be private, and I suggest it's easier to follow if the conditions are inverted:
```
	private static boolean isPhysicalMemoryTestExcluded() {
		/* Exclude on Linux aarch64 and x86-64
		 * https://github.com/eclipse/openj9/issues/12038
		 */
		String osname = System.getProperty("os.name");
		if ("Linux".equals(osname)) {
			String arch = System.getProperty("os.arch");
			if ("aarch64".equals(arch) || "amd64".equals(arch)) {
				return true;
			}
		}
		return false;
	}
```

pshipton commented:
Updated to detect when running in a Docker container on Linux rather than checking for hard coded platforms.

Tested via grinders, which don't detect a container as expected on the OpenJ9 machines. 
zlinux: https://ci.eclipse.org/openj9/view/Test/job/Grinder/1673
plinux: https://ci.eclipse.org/openj9/view/Test/job/Grinder/1674
xlinux: https://ci.eclipse.org/openj9/view/Test/job/Grinder/1675
aarch64: https://ci.eclipse.org/openj9/view/Test/job/Grinder/1676
osx: https://ci.eclipse.org/openj9/view/Test/job/Grinder/1677

Added an enum in the code, run one more grinder to validate.
https://ci.eclipse.org/openj9/view/Test/job/Grinder/1678

keithc-ca commented on test/functional/JLM_Tests/src/org/openj9/test/management/TestOperatingSystemMXBean.java:
```diff
@@ -64,6 +65,48 @@
+		}
+		String osname = System.getProperty("os.name");
+		if ("Linux".equals(osname)) {
+			try (FileInputStream env = new FileInputStream("/.dockerenv")) {
+				logger.info("/.dockerenv found");
+				inContainer = ContainerStatus.FOUND;
+				return true;
+			} catch (IOException e) {
+				// no "/.dockerenv" file, fall through
+			}
```
This can be simplified to `if (new File("/.dockerenv").exists()) { ...}`.

keithc-ca commented on test/functional/JLM_Tests/src/org/openj9/test/management/TestOperatingSystemMXBean.java:
```diff
@@ -64,6 +65,48 @@
+				// no "/.dockerenv" file, fall through
+			}
+			try (BufferedReader reader = new BufferedReader(new InputStreamReader(new FileInputStream("/proc/1/cgroup")))) {
+				String line;
+				while ((line = reader.readLine()) != null) {
+					if (line.contains("docker")) {
```
Suggested simplification:
```
		if (reader.lines().anyMatch(line -> line.contains("docker"))) {
			logger.info("Container found: " + line);
			inContainer = ContainerStatus.FOUND;
			return true;
		}
```

pshipton replied:
It doesn't work because `line` is used afterwards to log how the container was found, for easier debugging.

keithc-ca replied:
I missed that aspect. It could probably be reworked, but perhaps it's not worth it.

pshipton replied:
Updated it to use your pattern.

keithc-ca replied:
The lambda need not change if you use `findFirst()`:
```
	Optional<String> docker = reader.lines().filter(line -> line.contains("docker")).findFirst();
	if (docker.isPresent()) {
		logger.info("Container found: " + docker.get());
		inContainer = ContainerStatus.FOUND;
		return true;
	}
```

pshipton replied:
Updated.

keithc-ca commented on test/functional/JLM_Tests/src/org/openj9/test/management/TestOperatingSystemMXBean.java:
```diff
@@ -260,8 +303,12 @@ private static boolean test_PhysicalMemoryAPIs(com.ibm.lang.management.Operating
+			if (isPhysicalMemoryTestExcluded()) {
+				logger.info("Testing freeMemory <= totalMemory excluded on this platform");
+			} else {
+				Assert.assertFalse(freeMemory > totalMemory,
```
Can you improve the failure message so it includes `freeMemory` and `totalMemory`?

keithc-ca reviewed (commented):
Have you tested this in other container environments (e.g. `podman`)?

pshipton commented:
Updated to address comments. New grinder https://ci.eclipse.org/openj9/view/Test/job/Grinder/1681

andrew-m-leonard commented:
@pshipton just tried fix out at Adopt, and works greate on xLinux, pLinux and aarch64 docker nodes:
https://ci.adoptopenjdk.net/job/Test_openjdk16_j9_extended.functional_x86-64_linux/62/
https://ci.adoptopenjdk.net/job/Test_openjdk8_j9_extended.functional_ppc64le_linux/120/
https://ci.adoptopenjdk.net/view/Test_functional/job/Test_openjdk16_j9_extended.functional_aarch64_linux/61/

It would be useful to have this in the release branch if possible for next week so we avoid unnecessary test triage noise...

pshipton commented:
New grinder https://ci.eclipse.org/openj9/view/Test/job/Grinder/1686
